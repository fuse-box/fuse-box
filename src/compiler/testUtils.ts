import { generate } from './generator/generator';
import { ASTNode } from './interfaces/AST';
import { ImportType } from './interfaces/ImportType';
import { ITransformer } from './interfaces/ITransformer';
import { ITransformerRequireStatementCollection } from './interfaces/ITransformerRequireStatements';
import { parseTypeScript } from './parser';
import { createGlobalContext } from './program/GlobalContext';
import { transpileModule } from './program/transpileModule';
import { GlobalContextTransformer } from './transformers/GlobalContextTransformer';

function cleanupForTest(node) {
  delete node.loc;
  delete node.raw;
  delete node.range;
}
export function initCommonTransform(props: {
  jsx?: boolean;
  code: string;
  transformers: Array<ITransformer>;
  props?: any;
}) {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      cleanupForTest(statement.arguments[0]);
      cleanupForTest(statement);
      cleanupForTest(statement.callee);

      requireStatementCollection.push({ importType, statement });
    }
  }

  const ast = parseTypeScript(props.code, { jsx: props.jsx });

  // const ast = buntis.parseTSModule(props.code, {
  //   directives: true,
  //   jsx: props.jsx,
  //   next: true,
  //   loc: true,
  // });
  const tranformers = [GlobalContextTransformer().commonVisitors(props.props)];
  for (const t of props.transformers) {
    if (t.commonVisitors) {
      tranformers.push(t.commonVisitors({ ...props.props, onRequireCallExpression }));
    }
  }
  transpileModule({
    ast: ast as ASTNode,
    globalContext: createGlobalContext(),
    transformers: tranformers,
  });
  const res = generate(ast, {});

  return { code: res, requireStatementCollection };
}
