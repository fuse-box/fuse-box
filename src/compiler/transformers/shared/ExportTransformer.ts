import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import {
  ES_MODULE_EXPRESSION,
  createExports,
  createVariableDeclaration,
  isDefinedLocally,
} from '../../Visitor/helpers';
import { ASTNode, ASTType } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { ImportType } from '../../interfaces/ImportType';
import { GlobalContext } from '../../program/GlobalContext';

const IGNORED_DECLARATIONS = {
  [ASTType.InterfaceDeclaration]: 1,
  [ASTType.TypeAliasDeclaration]: 1,
};
function considerDecorators(node: ASTNode) {
  let statement: ASTNode = node.declaration;
  if (
    node.declaration.type === 'ClassDeclaration' &&
    node.declaration.decorators &&
    node.declaration.decorators.length
  ) {
    statement = createVariableDeclaration(node.declaration.id.name, node.declaration);
    // assigne decorators to the statement
    statement.decorators = node.declaration.decorators;
    statement.$fuse_decoratorForClassIdentifier = node.declaration.id.name;
    // remove them from the original declaration (in order for the decorators to be instered after)
    node.declaration.decorators = [];
  }
  return statement;
}
const INTERESTED_NODES = {
  ExportAllDeclaration: 1,
  ExportDefaultDeclaration: 1,
  ExportNamedDeclaration: 1,
};

export function ExportTransformer(): ITransformer {
  return {
    commonVisitors: props => {
      const definedLocallyProcessed: Record<string, number> = {};
      const compilerOptions = props.compilerOptions;
      let injectEsModuleStatement = compilerOptions.esModuleStatement === true;
      let esModuleStatementInjected = !injectEsModuleStatement;
      return {
        onEachNode: (visit: IVisit) => {
          const global = visit.globalContext as GlobalContext;

          if (global.exportAfterDeclaration) {
            const node = visit.node;
            const definedLocally = isDefinedLocally(node);

            if (definedLocally && !visit.node.$fuse_visited) {
              const newNodes = [];

              const response: IVisitorMod = {};

              for (const localVar of definedLocally) {
                // we don't want to add it multiple times
                if (definedLocallyProcessed[localVar.name]) break;
                definedLocallyProcessed[localVar.name] = 1;
                const targetAfter = global.exportAfterDeclaration[localVar.name];
                if (targetAfter) {
                  if (!response.appendToBody) response.appendToBody = [];
                  for (const item of targetAfter.targets) {
                    const ast = createExports(global.namespace, item, {
                      name: localVar.name,
                      type: 'Identifier',
                    });
                    response.appendToBody.push(ast);
                  }
                }
              }
              if (newNodes.length) {
                node.$fuse_visited = true;

                // console.log('new nodes', newNodes);
                // return { replaceWith: [node].concat(newNodes) };
                response.insertAfterThisNode = newNodes;
              }
              if (response.insertAfterThisNode || response.appendToBody) {
                // make sure exports.__eModule is injected
                if (!esModuleStatementInjected) {
                  esModuleStatementInjected = true;
                  response.prependToBody = [ES_MODULE_EXPRESSION];
                }
                return response;
              }
            }
          }
        },
        onTopLevelTraverse: (visit: IVisit) => {
          const node = visit.node;

          const global = visit.globalContext as GlobalContext;

          const type = node.type;
          if (!INTERESTED_NODES[type]) return;

          // remove export interface
          // export interface HelloWorld{}
          if (type === 'ExportNamedDeclaration') {
            if (node.declaration && IGNORED_DECLARATIONS[node.declaration.type]) {
              return { removeNode: true };
            }
          }

          /**
           * **************************************************************************
           * Export with source
           * export { one, two } from "./source"
           *
           * Needs to be simply replaced, no tracking involved
           */
          if (node.source) {
            const sourceVariable = global.getModuleName(node.source.value);

            // export * from "a"
            if (node.type === 'ExportAllDeclaration') {
              const reqExpression: ASTNode = {
                arguments: [
                  {
                    type: 'Literal',
                    value: node.source.value,
                  },
                ],
                callee: {
                  name: 'require',
                  type: 'Identifier',
                },
                type: 'CallExpression',
              };
              if (props.onRequireCallExpression) {
                props.onRequireCallExpression(ImportType.FROM, reqExpression);
              }
              const response: IVisitorMod = {
                replaceWith: {
                  arguments: [
                    {
                      name: 'exports',
                      type: 'Identifier',
                    },
                    reqExpression,
                  ],
                  callee: {
                    computed: false,
                    object: {
                      name: 'Object',
                      type: 'Identifier',
                    },
                    property: {
                      name: 'assign',
                      type: 'Identifier',
                    },
                    type: 'MemberExpression',
                  },
                  type: 'CallExpression',
                },
              };

              // make sure exports.__eModule is injected
              if (!esModuleStatementInjected) {
                esModuleStatementInjected = true;
                response.prependToBody = [ES_MODULE_EXPRESSION];
              }
              return response;
            }

            // export { foo } from "./bar"
            // creating
            //    var obj = require("module")
            const exportedNodes: Array<ASTNode> = [
              {
                declarations: [
                  {
                    id: {
                      name: sourceVariable,
                      type: 'Identifier',
                    },
                    init: {
                      arguments: [
                        {
                          type: 'Literal',
                          value: node.source.value,
                        },
                      ],
                      callee: {
                        name: 'require',
                        type: 'Identifier',
                      },
                      type: 'CallExpression',
                    },
                    type: 'VariableDeclarator',
                  },
                ],
                kind: 'var',
                type: 'VariableDeclaration',
              },
            ];

            if (type === 'ExportNamedDeclaration') {
              for (const specifier of node.specifiers) {
                exportedNodes.push(
                  createExports(global.namespace, specifier.exported.name, {
                    computed: false,
                    object: {
                      name: sourceVariable,
                      type: 'Identifier',
                    },
                    property: {
                      name: specifier.local.name,
                      type: 'Identifier',
                    },
                    type: 'MemberExpression',
                  }),
                );
              }
            }
            const response: IVisitorMod = { replaceWith: exportedNodes };
            // make sure exports.__eModule is injected
            if (!esModuleStatementInjected) {
              esModuleStatementInjected = true;
              response.prependToBody = [ES_MODULE_EXPRESSION];
            }
            return response;
          }

          /**
           * Handling export default declaratio
           * *********************************************************
           *    export default func
           */
          if (type === 'ExportDefaultDeclaration') {
            if (node.declaration) {
              if (node.declaration.type === ASTType.InterfaceDeclaration) {
                return { removeNode: true };
              }
              // Looking at this case:
              //      console.log(foo)
              //      export default function foo(){}
              if (node.declaration.type === 'FunctionDeclaration' || node.declaration.type === 'ClassDeclaration') {
                if (!node.declaration.id) {
                  node.declaration.id = { name: '__DefaultExport__', type: 'Identifier' };
                }

                // if there are decorators on the class we need to transform and assigne to a variable
                /**
                 * @dec
                 * export default class A {
                 *
                 * }
                 *
                 * Should be converted to:
                 *
                 * let A = class A {
                 * }
                 */
                let statement: ASTNode = considerDecorators(node);
                const response: IVisitorMod = {
                  replaceWith: [
                    statement,
                    createExports(global.namespace, 'default', {
                      name: node.declaration.id.name,
                      type: 'Identifier',
                    }),
                  ],
                };
                // make sure exports.__eModule is injected
                if (!esModuleStatementInjected) {
                  esModuleStatementInjected = true;
                  response.prependToBody = [ES_MODULE_EXPRESSION];
                }
                return response;
              }
              const response: IVisitorMod = {
                replaceWith: createExports(global.namespace, 'default', node.declaration),
              };
              // make sure exports.__eModule is injected
              if (!esModuleStatementInjected) {
                esModuleStatementInjected = true;
                response.prependToBody = [ES_MODULE_EXPRESSION];
              }
              return response;
            }
          }

          if (type === 'ExportNamedDeclaration') {
            /**
             * ******************************************************************************
             * CASE with variable exports
             *
             *      export { name1, name2 }
             *
             * Will be "soft" replaced if those defined earlier
             * Then if it's not defined we need to wait for those (they might come later)
             */

            const newNodes = [];

            if (node.specifiers && node.specifiers.length) {
              for (const specifier of node.specifiers) {
                // if the variable present in the scope...
                // or picked up by the identifierReplacement ( we can leave that for GlobalContextTranformer)
                if (
                  (visit.scope && visit.scope.locals[specifier.local.name]) ||
                  global.identifierReplacement[specifier.local.name]
                ) {
                  newNodes.push(
                    createExports(global.namespace, specifier.exported.name, {
                      name: specifier.local.name,
                      type: 'Identifier',
                    }),
                  );
                } else {
                  // if none was decleared, it must be declared earlier (so we check if later with onEachNode in this transformer )
                  let exportAfterDeclaration = global.exportAfterDeclaration;
                  if (!exportAfterDeclaration) {
                    global.exportAfterDeclaration = exportAfterDeclaration = {};
                  }

                  if (!exportAfterDeclaration[specifier.local.name]) {
                    exportAfterDeclaration[specifier.local.name] = { targets: [] };
                  }
                  // it can have multiple exports of the same variable. For example
                  // export {foo, foo as foo1}

                  exportAfterDeclaration[specifier.local.name].targets.push(specifier.exported.name);
                }
              }
              const response: IVisitorMod = { replaceWith: newNodes };
              // make sure exports.__eModule is injected
              if (!esModuleStatementInjected) {
                esModuleStatementInjected = true;
                response.prependToBody = [ES_MODULE_EXPRESSION];
              }
              return response;
            }
            /**
             * ************************************************************************
             * Export objects directly
             *
             *  export function hello(){}
             *  export class Bar(){}
             */
            if (node.declaration) {
              if (node.declaration.id && node.declaration.id.name) {
                const statement = considerDecorators(node);
                return {
                  replaceWith: [
                    statement,
                    createExports(global.namespace, node.declaration.id.name, {
                      name: node.declaration.id.name,
                      type: 'Identifier',
                    }),
                  ],
                };
              }
              if (node.declaration.type === 'VariableDeclaration') {
                if (node.declaration.declare) {
                  return { ignoreChildren: true, removeNode: true };
                }
                /**
                 * ******************************************************************
                 * Exports defined constants
                 *
                 *    export const foo = 1, bar = 3;
                 */
                if (node.declaration.declarations) {
                  let newNodes = [];
                  for (const declaration of node.declaration.declarations) {
                    if (!declaration.declare) {
                      global.identifierReplacement[declaration.id.name] = {
                        first: global.namespace,
                        second: declaration.id.name,
                      };

                      if (declaration.init) {
                        // check if there is INIT
                        // we night have something like this:
                        //    export var FooBar;
                        newNodes.push(createExports(global.namespace, declaration.id.name, declaration.init));
                      }
                    }
                  }
                  const response: IVisitorMod = { replaceWith: newNodes };

                  // make sure exports.__eModule is injected
                  if (!esModuleStatementInjected) {
                    esModuleStatementInjected = true;
                    response.prependToBody = [ES_MODULE_EXPRESSION];
                  }
                  return response;
                }
              }
            }
          }
        },
      };
    },
  };
}
