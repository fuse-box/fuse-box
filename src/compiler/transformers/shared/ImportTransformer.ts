import * as path from 'path';
import { GlobalContext } from '../../program/GlobalContext';
import { ITransformer } from '../../program/transpileModule';
import { createRequireStatement } from '../../Visitor/helpers';
import { IVisit } from '../../Visitor/Visitor';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import { ImportType } from '../../interfaces/ImportType';

// FIX SHORTHANDS
export function ImportTransformer(options?: ITransformerSharedOptions): ITransformer {
  options = options || {};
  return {
    onTopLevelTraverse: (visit: IVisit) => {
      const node = visit.node;
      const global = visit.globalContext as GlobalContext;

      if (node.type === 'ImportDeclaration') {
        // converts "./foo/bar.hello.js" to foo_bar_hello_js_1 (1:1 like typescript does)
        const variable = path.basename(node.source.value).replace(/\.|-/, '_') + '_' + global.getNextIndex();

        node.specifiers.forEach(specifier => {
          if (specifier.type === 'ImportSpecifier') {
            global.identifierReplacement[specifier.local.name] = {
              first: variable,
              second: specifier.imported.name,
            };
          } else if (specifier.type === 'ImportDefaultSpecifier') {
            global.identifierReplacement[specifier.local.name] = {
              first: variable,
              second: 'default',
            };
          } else if (specifier.type === 'ImportNamespaceSpecifier') {
            // only if we have more than one specifier
            // for instance
            // i=mport MySuperClass, * as everything from "module-name";

            // in every other case like:
            // import * as everything from "module-name";
            // we don't need to do anything, since the variables should match

            global.identifierReplacement[specifier.local.name] = {
              first: variable,
            };

            //if (node.specifiers.length > 1) afterStatement = defineVariable(variable, specifier.local.name);
          }
        });

        return {
          onComplete: () => {
            const reqStatement = createRequireStatement(node.source.value, node.specifiers.length && variable);
            // when everything is finished we need to check if those variables have been used at all
            // they were all unused we need remove the require/import statement at all

            // assuming we have no specififers and this import HAS side effects
            // e.g import "./module"
            const parent = visit.parent;
            const property = visit.property;
            if (node.specifiers.length === 0) {
              const index = parent[property].indexOf(node);
              if (index > -1) {
                if (options.onRequireCallExpression) {
                  options.onRequireCallExpression(ImportType.RAW_IMPORT, reqStatement.reqStatement);
                }
                parent[property].splice(index, 1, reqStatement.statement);
              }
              return;
            }

            let atLeastOneInUse = false;
            for (const specifier of node.specifiers) {
              const traced = global.identifierReplacement[specifier.local.name];
              if (traced && traced.inUse) {
                atLeastOneInUse = true;
                break; // we just need to know if we need to keep the node
              }
            }
            // doing a manual replace
            if (atLeastOneInUse) {
              const index = parent[property].indexOf(node);
              if (index > -1) {
                parent[property].splice(index, 1, reqStatement.statement);
                if (options.onRequireCallExpression) {
                  options.onRequireCallExpression(ImportType.FROM, reqStatement.reqStatement);
                }
              }
            } else {
              const index = parent[property].indexOf(node);
              if (index > -1) parent[property].splice(index, 1);
            }
          },
        };
      }
    },
  };
}
