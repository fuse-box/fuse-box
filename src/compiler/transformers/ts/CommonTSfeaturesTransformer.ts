import { ISchema } from '../../core/nodeSchema';
import { ASTType } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';

const FUNC_EXPRESSIONS = { FunctionDeclaration: 1, FunctionExpression: 1 };
// to test: function maybeUnwrapEmpty<T>(value: T[]): T[];
// to test: (oi as any).foo

export function CommonTSfeaturesTransformer(): ITransformer {
  return {
    target: { type: 'ts' },
    commonVisitors: props => {
      return {
        onEach: (schema: ISchema) => {
          const { node, parent, property } = schema;

          // handle "this" typings
          // e.g function hey(this: number, a) {}
          if (
            parent &&
            property === 'params' &&
            FUNC_EXPRESSIONS[parent.type] &&
            node.type === 'Identifier' &&
            node.name === 'this'
          ) {
            return schema.remove();
          }

          if (node.declare) {
            return schema.remove();
          }

          if (node.type === ASTType.TypeAssertion || node.type === ASTType.NonNullExpression) {
            return schema.replace([node.expression]);
          }
          // EmptyBodyFunctionExpression is a specific buntis key
          if (
            node.type === 'MethodDefinition' &&
            node.value &&
            (node.value.type === 'EmptyBodyFunctionExpression' || !node.value.body)
          ) {
            return schema.remove();
          }
          switch (node.type) {
            case 'ClassProperty':
            case ASTType.AbstractClassProperty:
            case ASTType.AbstractMethodDefinition:
            case ASTType.DeclareFunction:
            case ASTType.IndexSignature:
            case ASTType.InterfaceDeclaration:
            case ASTType.TypeAliasDeclaration:
              return schema.remove();
            case 'ExportNamedDeclaration':
              const decl = node.declaration;
              if (decl) {
                if (decl.declare || decl.type === ASTType.InterfaceDeclaration) {
                  return schema.remove();
                }
              }
              break;
            case ASTType.AsExpression:
              return schema.replace([node.expression]);

            case ASTType.ParameterProperty:
              return schema.replace([node.parameter]);
          }
        },
        onProgramBody: (schema: ISchema) => {
          if (schema.node.declare) {
            return schema.remove();
          }
        },
      };
    },
  };
}
