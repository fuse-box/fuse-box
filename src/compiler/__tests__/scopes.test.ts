import { initCommonTransform } from '../testUtils';
import { ExportTransformer } from '../transformers/shared/ExportTransformer';
import { ImportTransformer } from '../transformers/shared/ImportTransformer';
import { ClassConstructorPropertyTransformer } from '../transformers/ts/ClassConstructorPropertyTransformer';
import { CommonTSfeaturesTransformer } from '../transformers/ts/CommonTSfeaturesTransformer';

const testTranspile = (props: { code: string; fileName?: string; target?: string }) => {
  return initCommonTransform({
    code: props.code,
    transformers: [
      ImportTransformer(),
      ExportTransformer(),
      CommonTSfeaturesTransformer(),
      ClassConstructorPropertyTransformer(),
    ],
  });
};

describe('scope test', () => {
  it('option 1', () => {
    const result = testTranspile({
      code: `
      import { hey } from 'oi';
      function hey(){}
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('option 2', () => {
    const result = testTranspile({
      code: `
      import { hey } from 'oi';
      function hey(){}
      console.log(hey);
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('variation 3', () => {
    const result = testTranspile({
      code: `
      import { hey } from 'oi';
      function foo(hey){}
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('variation 4', () => {
    const result = testTranspile({
      code: `
      import { hey } from 'oi';
      const hey = (hey) => {}
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('variation 5', () => {
    const result = testTranspile({
      code: `
      import { hey } from 'oi';
      const hey = (hey) => {
        console.log(hey)
      }
      `,
    });
    expect(result.code).toMatchSnapshot();
  });
  it('variation 6', () => {
    const result = testTranspile({
      code: `
      import { hey } from 'oi';
      console.log(hey);
      const hey = (hey) => {
        console.log(hey)
      }
      `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('variation 7', () => {
    const result = testTranspile({
      code: `
      import { Hey } from 'oi';
      class Hey {
        foo(hey){ console.log(hey)}
      }
      `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('variation with class After', () => {
    const result = testTranspile({
      code: `
      import { Hey } from 'oi';
      console.log(Hey);
      class Hey {
        foo(hey){ console.log(hey)}
      }
      new Hey()
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('variation with arrow expression', () => {
    const result = testTranspile({
      code: `
      export const a = () => {};
      console.log(() => a);
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('variation compute properties', () => {
    const result = testTranspile({
      code: `
      export const a = () => {};
      const b = {[a] : 1}
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('variation normal prop', () => {
    const result = testTranspile({
      code: `
      export const a = () => {};
      const b = {a : a}
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('Array pattern', () => {
    const result = testTranspile({
      code: `
      import foo from "oi";
      function one(props){
        const [foo] = props;
        console.log(foo)
      }
      console.log(foo)
      `,
    });

    expect(result.code).toMatchSnapshot();
  });
  it('Object pattern', () => {
    const result = testTranspile({
      code: `
      import foo from "oi";
      function one(props){
        const {foo} = props;
        console.log(foo)
      }
      console.log(foo)
      `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('Object pattern with spread', () => {
    const result = testTranspile({
      code: `
      import foo from "oi";
      function one(props){
        const {foo, ...rest} = props;
        console.log(foo)
      }
      console.log(foo)
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('Object respect spread in Object', () => {
    const result = testTranspile({
      code: `
      import {foo, rest} from "oi";
      function one(props){
        const {foo, ...rest} = props;
        console.log(rest)

      }
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('Object respect spread in Array', () => {
    const result = testTranspile({
      code: `
      import {foo, rest} from "oi";
      function one(props){
        const [foo, ...rest] = props;
        console.log(rest)
      }
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should work with nulll', () => {
    const result = testTranspile({
      code: `
      import {foo, rest} from "oi";
      function one(props){
        const [, foo, ...rest] = props;
        console.log(rest)
      }
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should handle hoisted variable', () => {
    const result = testTranspile({
      code: `
      export var Foo = (function() {
        console.log(Foo);
        function Foo() {}
      })();
      `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should handle hoisted variable deep', () => {
    const result = testTranspile({
      code: `
      export var Foo = (function() {
        function Bar(){
          console.log(Foo);
        }
        function Foo() {}
      })();
      `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should handle hoisted variable 2', () => {
    const result = testTranspile({
      code: `
      function some() {}
      export var Disposable = (function() {
        function Disposable() {}
        return Disposable;
      })();
      console.log(Disposable);

      `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should handle hoisted var inside', () => {
    const result = testTranspile({
      code: `
      import { execute } from 'stuff';
      function hello() {}
      function some() {
        function execute(){}
      }
      execute();

      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should handle hoisted var inside 2', () => {
    const result = testTranspile({
      code: `
      import { execute } from 'stuff';

      function some() {
        function execute(){}
      }
      execute();

      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should handle hoisted var inside 3', () => {
    const result = testTranspile({
      code: `
      import { execute } from 'stuff';
      function execute(){}
      execute();

      `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should handle hoisted var inside 4', () => {
    const result = testTranspile({
      code: `
      import { isFunction, noop } from '../utils/js_utils';
      export var Disposable = (function() {
        function Disposable(action) {}

        return Disposable;
      })();
      console.log(Disposable);
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should handle hoisted var inside 5', () => {
    const result = testTranspile({
      code: `
      export var Disposable = (function() {
        function Disposable(action) {}

        return Disposable;
      })();
      console.log(Disposable);
      `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should handle hoisted var inside 6', () => {
    const result = testTranspile({
      code: `
      import { isFunction, noop } from '../utils/js_utils';
      export var Disposable = (function() {
        function Disposable(action) {
          console.log(isFunction)
        }

        return Disposable;
      })();
      console.log(Disposable);
      `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should handle hoisted var Buffer', () => {
    const result = testTranspile({
      code: `
      'use strict'

      var base64 = require('base64-js');
      var ieee754 = require('ieee754');

      exports.Buffer = Buffer;
      function Buffer(){}
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should handle hoisted var Buffer 2', () => {
    const result = testTranspile({
      code: `
      'use strict'
      function two(){}
      function one(){}
      console.log(Buffer)
      function Buffer(){}
      function three(){}
      `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should handle import buffer', () => {
    const result = testTranspile({
      code: `
      'use strict'
      import Buffer from "other"
      function two(){}
      function one(){}
      console.log(Buffer)

      function three(){}
      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should trace down an import 1', () => {
    const result = testTranspile({
      code: `
      import zenObservable from 'zen-observable';
      var Observable = zenObservable;

      `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should handle multiple scenarious at one', () => {
    const result = testTranspile({
      code: `
      import { execute } from 'shit';
      import foo from 'oi';
      import { Hey } from 'hey';

      function hello() {}
      function some() {
        const execute = {};
      }
      execute();

      function one(props) {
        const { foo, ...rest } = props;
        console.log(foo);
      }
      console.log(foo);

      console.log(Hey);
      class Hey {
        foo(hey) {
          console.log(hey);
        }
      }
      new Hey();

      function some() {}
      export var Disposable = (function() {
        function Disposable() {}
        return Disposable;
      })();
      console.log(Disposable);

      `,
    });
    expect(result.code).toMatchSnapshot();
  });

  describe('Spread', () => {
    it('spread and defined var', () => {
      const result = testTranspile({
        code: `
        import oi from "./oi"
        function foo(){
          function boo(){
            const {foo : { bar : [ oi ] }} = {}
            function another(){
              console.log(oi);
            }
          }
          console.log(oi);
        }


      `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('Export spread 1', () => {
      const result = testTranspile({
        code: `
        const [foo, bar] = []
        export {foo, bar}
      `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('Export spread 2', () => {
      const result = testTranspile({
        code: `
        const {foo, bar} = {}
        export {foo, bar}
      `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('Export spread 3', () => {
      const result = testTranspile({
        code: `
        const {foo : { bar }} = {}
        export { bar }
      `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('Export spread 4 (deep)', () => {
      const result = testTranspile({
        code: `
        const {foo : { bar : [ oi ] }} = {}
        export { oi }
      `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('Export spread 4 (deep) skip key', () => {
      const result = testTranspile({
        code: `
        const {foo : { bar : [ oi ] }} = {}
        export { oi, bar }
      `,
      });
      expect(result.code).toMatchSnapshot();
    });
  });

  describe('classes', () => {
    it('should respect method prop', () => {
      const result = testTranspile({
        code: `
        import { bar } from "bar";
        class Foo {
          method_1(bar){
            console.log(bar)
          }
        }
      `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should respect constructor prop', () => {
      const result = testTranspile({
        code: `
        import { bar } from "bar";
        class Foo {
          constructor(bar){
            console.log(bar)
          }
        }
      `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should respect import qualifer id ref', () => {
      const result = testTranspile({
        code: `
        import {Foo} from 'oi';
        import Bar = Foo;
      `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should respect import qualifer id ref deep', () => {
      const result = testTranspile({
        code: `
        import {Foo} from 'oi';
        import Bar = Foo.baz;
      `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should respect constructor prop with public accessibility', () => {
      const result = testTranspile({
        code: `
        import { bar } from "bar";
        class Foo {
          constructor(public bar){
            console.log(bar)
          }
        }
      `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should respect catch clause', () => {
      const result = testTranspile({
        code: `
        import someException from 'someException';
        function oi() {}
        function hey() {
          try {
            oi();
          } catch (someException) {
            console.log(someException);
          }
        }

      `,
      });

      expect(result.code).toMatchSnapshot();
    });
  });
});
