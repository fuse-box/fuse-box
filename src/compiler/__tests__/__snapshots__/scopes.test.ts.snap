// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scope test Array pattern 1`] = `
"var oi_1 = require(\\"oi\\");
function one(props) {
  const [foo] = props;
  console.log(foo);
}
console.log(oi_1.default);
"
`;

exports[`scope test Object pattern 1`] = `
"var oi_1 = require(\\"oi\\");
function one(props) {
  const {foo} = props;
  console.log(foo);
}
console.log(oi_1.default);
"
`;

exports[`scope test Object pattern with spread 1`] = `
"var oi_1 = require(\\"oi\\");
function one(props) {
  const {foo, ...rest} = props;
  console.log(foo);
}
console.log(oi_1.default);
"
`;

exports[`scope test Object respect spread in Array 1`] = `
"function one(props) {
  const [foo, ...rest] = props;
  console.log(rest);
}
"
`;

exports[`scope test Object respect spread in Object 1`] = `
"function one(props) {
  const {foo, ...rest} = props;
  console.log(rest);
}
"
`;

exports[`scope test option 1 1`] = `
"function hey() {}
"
`;

exports[`scope test option 2 1`] = `
"function hey() {}
console.log(hey);
"
`;

exports[`scope test should work with nulll 1`] = `
"function one(props) {
  const [, foo, ...rest] = props;
  console.log(rest);
}
"
`;

exports[`scope test variation 3 1`] = `
"function foo(hey) {}
"
`;

exports[`scope test variation 4 1`] = `
"const hey = hey => {};
"
`;

exports[`scope test variation 5 1`] = `
"const hey = hey => {
  console.log(hey);
};
"
`;

exports[`scope test variation 6 1`] = `
"var oi_1 = require(\\"oi\\");
console.log(oi_1.hey);
const hey = hey => {
  console.log(hey);
};
"
`;

exports[`scope test variation 7 1`] = `
"class Hey {
  foo(hey) {
    console.log(hey);
  }
}
"
`;

exports[`scope test variation compute properties 1`] = `
"exports.a = () => {};
const b = {
  [exports.a]: 1
};
"
`;

exports[`scope test variation normal prop 1`] = `
"exports.a = () => {};
const b = {
  a: exports.a
};
"
`;

exports[`scope test variation with arrow expression 1`] = `
"exports.a = () => {};
console.log(() => exports.a);
"
`;

exports[`scope test variation with class After 1`] = `
"var oi_1 = require(\\"oi\\");
console.log(oi_1.Hey);
class Hey {
  foo(hey) {
    console.log(hey);
  }
}
new Hey();
"
`;
